File: src/Card.java
public class Card {
    private Rank rank;
    private Suit suit;

    public Card(Suit suit, Rank rank) {
        this.suit = suit;
        this.rank = rank;
    }
    // Add a constructor that takes a String consisting of two characters as an argument,
    // with the suit as first character and rank as second character
    // examples: "D2"= DIAMOND-TWO, "HA"= HEART-ACE, "ST"= SPADE-TEN, "CK"= CLUB-KING
    public Card(String cardDesc) {
        cardDesc = cardDesc.toUpperCase();  // just in case the input is lowercase
        if (cardDesc.length() != 2) {
            throw new IllegalArgumentException("Invalid card description: " + cardDesc);
        }
        switch (cardDesc.charAt(0)) {
            case 'C': this.suit = Suit.CLUB; break;
            case 'D': this.suit = Suit.DIAMOND; break;
            case 'H': this.suit = Suit.HEART; break;
            case 'S': this.suit = Suit.SPADE; break;
            default:
                throw new IllegalArgumentException("Invalid suit: " + cardDesc.charAt(0));
        }
        switch (cardDesc.charAt(1)) {
            case '2': this.rank = Rank.TWO; break;
            case '3': this.rank = Rank.THREE; break;
            case '4': this.rank = Rank.FOUR; break;
            case '5': this.rank = Rank.FIVE; break;
            case '6': this.rank = Rank.SIX; break;
            case '7': this.rank = Rank.SEVEN; break;
            case '8': this.rank = Rank.EIGHT; break;
            case '9': this.rank = Rank.NINE; break;
            case 'T': this.rank = Rank.TEN; break;
            case 'J': this.rank = Rank.JACK; break;
            case 'Q': this.rank = Rank.QUEEN; break;
            case 'K': this.rank = Rank.KING; break;
            case 'A': this.rank = Rank.ACE; break;
            default:
                throw new IllegalArgumentException("Invalid rank: " + cardDesc.charAt(1));
        }
    }

    public Rank getRank() {
        return rank;
    }

    public Suit getSuit() {
        return suit;
    }

    @Override
    public String toString() {
        return suit + "-" + rank;
    }
}

File: src/golang/poker/card.go
package poker

import (
	"fmt"
	"strings"
)

// Card represents a playing card with a suit and rank
type Card struct {
	Suit Suit
	Rank Rank
}

// NewCard creates a new Card from Suit and Rank
func NewCard(suit Suit, rank Rank) Card {
	return Card{Suit: suit, Rank: rank}
}

// NewCardFromString creates a new Card from a two-character string
func NewCardFromString(cardDesc string) (Card, error) {
	cardDesc = strings.ToUpper(cardDesc)
	if len(cardDesc) != 2 {
		return Card{}, fmt.Errorf("invalid card description: %s", cardDesc)
	}

	var suit Suit
	switch cardDesc[0] {
	case 'C':
		suit = CLUB
	case 'D':
		suit = DIAMOND
	case 'H':
		suit = HEART
	case 'S':
		suit = SPADE
	default:
		return Card{}, fmt.Errorf("invalid suit: %c", cardDesc[0])
	}

	var rank Rank
	switch cardDesc[1] {
	case '2':
		rank = TWO
	case '3':
		rank = THREE
	case '4':
		rank = FOUR
	case '5':
		rank = FIVE
	case '6':
		rank = SIX
	case '7':
		rank = SEVEN
	case '8':
		rank = EIGHT
	case '9':
		rank = NINE
	case 'T':
		rank = TEN
	case 'J':
		rank = JACK
	case 'Q':
		rank = QUEEN
	case 'K':
		rank = KING
	case 'A':
		rank = ACE
	default:
		return Card{}, fmt.Errorf("invalid rank: %c", cardDesc[1])
	}

	return Card{Suit: suit, Rank: rank}, nil
}

func (c Card) String() string {
	return fmt.Sprintf("%s-%s", c.Suit, c.Rank)
}
